elasticdl/python/common/tensor.py

```diff
@@ -0,0 +1,116 @@
+import numpy as np
+import tensorflow as tf
+
+from elasticdl.proto import elasticdl_pb2
+from elasticdl.python.common.dtypes import (
+    dtype_numpy_to_tensor,
+    dtype_tensor_to_numpy,
+)
+
+
+def serialize_tensor(tensor, tensor_pb):
+    """Convert Tensor to Tensor PB"""
+    dtype = dtype_numpy_to_tensor(tensor.values.dtype)
+    if not dtype:
+        raise ValueError(
+            "Dtype of ndarray %s is not supported", tensor.values.dtype
+        )
+    tensor_pb.dtype = dtype
+    tensor_pb.dim.extend(tensor.values.shape)
+    tensor_pb.content = tensor.values.tobytes()
+    if tensor.is_indexed_slices():
+        tensor_pb.indices.extend(tensor.indices)
+    if tensor.name:
+        tensor_pb.name = tensor.name
+
+
+def deserialize_tensor_pb(tensor_pb, tensor):
+    """Create a Tensor instance from Tensor proto message.
+
+    Note: upon return, the input tensor message is reset and underlying buffer
+    passed to the returned ndarray.
+    """
+
+    if not tensor_pb.dim:
+        raise ValueError("Tensor PB has no dim defined")
+
+    dtype = dtype_tensor_to_numpy(tensor_pb.dtype)
+    # Check that the buffer size agrees with dimensions.
+    size = dtype.itemsize
+    for d in tensor_pb.dim:
+        size *= d
+    if size != len(tensor_pb.content):
+        raise ValueError(
+            "Tensor PB size mismatch, dim: %s, len(content): %d",
+            tensor_pb.dim,
+            len(tensor_pb.content),
+        )
+    tensor = Tensor()
+    tensor.set(
+        np.ndarray(shape=tensor_pb.dim, dtype=dtype, buffer=tensor_pb.content),
+        tensor_pb.indices,
+        tensor_pb.name,
+    )
+    tensor_pb.Clear()
```

Comment:
Do we need to move this to a finally block?

Reply:
I think that we should move this line.
tensor_pb is the input of this function. The caller usually not except the tensor_pb is cleared when using this function. The caller is responsible for clearing tensor_pb at some place.

Reply:
@QiJune Good point. In fact, I clear `tensor_pb` because `common.ndarray.tensor_to_ndarray` does, and I made some mistakes because I forget this function will clear `tensor_pb`... 

For me, clearing `tensor_pb` in this function is confusing.

Reply:
In common, tensor_pb is a local variable, it will be deleted by Python Interpreter at some time automatically.
If we really want to clear it right now, we should call `tensor_pb.Clear()` explicitly at some place, but not put it inside a deserialize function.

For example:

```python
deserialize_tensor_pb(tensor_pb, tensor)
tensor_pb.Clear()
```

Anyway, this line should be removed from this function.

Reply:
Yes, I agree with moving this line out.